core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+		trace!(target: "babe", "Verifying something");
+		let mut inherent_data = self.inherent_data_providers.create_inherent_data().map_err(String::from)?;
+		let (_, slot_now) = BabeSlotCompatible::extract_timestamp_and_slot(&inherent_data)
+			.map_err(|e| format!("Could not extract timestamp and slot: {:?}", e))?;
+		let hash = header.hash();
+		let parent_hash = *header.parent_hash();
+		let authorities = self.authorities(&BlockId::Hash(parent_hash))
+			.map_err(|e| format!("Could not fetch authorities at {:?}: {:?}", parent_hash, e))?;
+
+		let extra_verification = self.extra.verify(
+			&header,
+			body.as_ref().map(|x| &x[..]),
+		);
+
+		// we add one to allow for some small drift.
+		// FIXME #1019 in the future, alter this queue to allow deferring of headers
+		let checked_header = check_header::<B>(
+			slot_now + 1,
+			header,
+			hash,
+			&authorities[..],
+			self.threshold,
+		)?;
+		match checked_header {
+			CheckedHeader::Checked(pre_header, seal) => {
+				let BabeSeal { slot_num, .. } = seal.as_babe_seal()
+					.expect("check_header always returns a seal digest item; qed");
+
+				// if the body is passed through, we need to use the runtime
+				// to check that the internally-set timestamp in the inherents
+				// actually matches the slot set in the seal.
+				if let Some(inner_body) = body.take() {
+					inherent_data.babe_replace_inherent_data(slot_num);
+					let block = B::new(pre_header.clone(), inner_body);
+
+					self.check_inherents(
+						block.clone(),
+						BlockId::Hash(parent_hash),
+						inherent_data,
+					)?;
+
+					let (_, inner_body) = block.deconstruct();
+					body = Some(inner_body);
+				}
+
+				trace!(target: "babe", "Checked {:?}; importing.", pre_header);
+				telemetry!(CONSENSUS_TRACE; "babe.checked_and_importing"; "pre_header" => ?pre_header);
+
+				extra_verification.into_future().wait()?;
+
+				let import_block = ImportBlock {
+					origin,
+					header: pre_header,
+					post_digests: vec![seal],
+					body,
+					finalized: false,
+					justification,
+					auxiliary: Vec::new(),
+					fork_choice: ForkChoiceStrategy::LongestChain,
+				};
+
+				// FIXME #1019 extract authorities
+				Ok((import_block, None))
+			}
+			CheckedHeader::Deferred(a, b) => {
+				debug!(target: "babe", "Checking {:?} failed; {:?}, {:?}.", hash, a, b);
+				telemetry!(CONSENSUS_DEBUG; "babe.header_too_far_in_future";
+					"hash" => ?hash, "a" => ?a, "b" => ?b
+				);
+				Err(format!("Header {:?} rejected: too far in the future", hash))
+			}
+		}
+	}
+}
+
+impl<B, C, E> Authorities<B> for BabeVerifier<C, E> where
+	B: Block,
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+{
+	type Error = ConsensusError;
+
+	fn authorities(&self, at: &BlockId<B>) -> Result<Vec<AuthorityIdFor<B>>, Self::Error> {
+		authorities(self.client.as_ref(), at)
+	}
+}
+
+fn authorities<B, C>(client: &C, at: &BlockId<B>) -> Result<Vec<AuthorityIdFor<B>>, ConsensusError> where
```

Comment:
As mentioned before maybe return a `HashSet` here

Reply:
That would not help, since we return an owned `Vec` here, which is already O(n) to construct.